### PR TITLE
descriptors: Introduce sortedmulti descriptor

### PR DESCRIPTION
**Pull request by achow101**
Adds a `sortedmulti()` descriptor as mentioned in https://github.com/bitcoin/bitcoin/pull/17023#issuecomment-537596416.

`sortedmulti()` works in the same way as `multi` does but sorts the pubkeys in the resulting scripts in lexicographic order as described in [BIP67](https://github.com/bitcoin/bips/blob/master/bip-0067.mediawiki). Note that this does not add support for BIP67 nor is BIP67 fully supported by this descriptor (which is why it is not named `multi67()`) as it does not require compressed pubkeys.

Tests from BIP67 were added and documentation was updated.

### CODE REVIEW

File: src/script/descriptor.cpp [hunk 1]
```diff
@@ -593,15 +593,23 @@ class ComboDescriptor final : public DescriptorImpl
     ComboDescriptor(std::unique_ptr<PubkeyProvider> prov) : DescriptorImpl(Singleton(std::move(prov)), {}, "combo") {}
 };
 
-/** A parsed multi(...) descriptor. */
+/** A parsed multi(...) or sortedmulti(...) descriptor */
 class MultisigDescriptor final : public DescriptorImpl
 {
     const int m_threshold;
+    bool m_sorted;
```
Comment (promag) [inline src/script/descriptor.cpp]: Also const?

File: src/script/descriptor.cpp [hunk 2]
```diff
@@ -827,7 +836,10 @@ std::unique_ptr<DescriptorImpl> ParseScript(Span<const char>& sp, ParseScriptCon
         error = "Cannot have combo in non-top level";
         return nullptr;
     }
-    if (Func("multi", expr)) {
+    if (Func("sortedmulti", expr)) {
+        sorted_multi = true;
+    }
+    if (Func("multi", expr) || sorted_multi) {
```
Comment (promag) [inline src/script/descriptor.cpp]: Swap operands?

**Comment by Sjors** [comment]
Concept ACK

@peter-conalgo does this match the BIP67 implementation in ColdCard? (assuming compressed public keys). E.g. https://github.com/Coldcard/firmware/blob/master/testing/test_multisig.py#L232-L255 and https://github.com/Coldcard/firmware/blob/master/shared/multisig.py#L412-L414

File: test/functional/rpc_createmultisig.py [hunk 3]
```diff
@@ -72,6 +75,14 @@ def run_test(self):
             assert_equal(legacy_addr, node0.addmultisigaddress(2, keys, '', 'bech32')['address'])
             assert_equal(legacy_addr, node0.addmultisigaddress(2, keys, '', 'p2sh-segwit')['address'])
 
+        self.log.info('Testing sortedmulti descriptors with BIP 67 test vectors')
+        with open(os.path.join(os.path.dirname(os.path.realpath(__file__)), 'data/rpc_bip67.json'), encoding='utf-8') as f:
+            vectors = json.load(f)
+
+        for t in vectors:
+            key_str = ','.join(t['keys'])
+            self.nodes[0].deriveaddresses(descsum_create('sh(sortedmulti(2,' + key_str + '))'))[0] == t['address']
```
Comment (Sjors) [inline test/functional/rpc_createmultisig.py]: This just returns `false` without failing the test.  In fact all these addresses are incorrect; you either have to run the node in mainnet (like `feature_config_args`), or replace the test vectors for regtest. I manually verified with a mainnet node that the vectors do pass.

**Comment by Sjors** [comment]
I tested with https://github.com/cryptoadvance/specter-desktop/pull/8 that ColdCard simulator no longer complains about a change address that it complained about before.

Code review ACK d5a925b except for the functional test issue.

**Review by Sjors — CHANGES_REQUESTED** [review]
I tested with https://github.com/cryptoadvance/specter-desktop/pull/8 that ColdCard simulator no longer complains about a change address that it complained about before.

Code review ACK d5a925b except for the functional test issue.

File: test/functional/rpc_createmultisig.py (hunk 3 above)
Comment (achow101) [reply test/functional/rpc_createmultisig.py]: Whoops. Fixed. Changed the addresses to regtest ones. You can check they still map to the same scriptPubKey using `getaddressinfo` and `decodescript`.

**Comment by Sjors** [comment]
ACK a6dd441cb2d8325dd968897776e7f8389cda359c

File: doc/descriptors.md [hunk 4]
```diff
@@ -37,12 +38,14 @@ Output descriptors currently support:
 - `sh(wsh(pkh(02e493dbf1c10d80f3581e4904930b1404cc6c13900ee0758474fa94abe8c4cd13)))` describes an (overly complicated) P2SH-P2WSH-P2PKH output with the specified public key.
 - `multi(1,022f8bde4d1a07209355b4a7250a5c5128e88b84bddc619ab7cba8d569b240efe4,025cbdf0646e5db4eaa398f365f2ea7a0e3d419b7e0330e39ce92bddedcac4f9bc)` describes a bare *1-of-2* multisig output with keys in the specified order.
 - `sh(multi(2,022f01e5e15cca351daff3843fb70f3c2f0a1bdd05e5af888a67784ef3e10a2a01,03acd484e2f0c7f65309ad178a9f559abde09796974c57e714c35f110dfc27ccbe))` describes a P2SH *2-of-2* multisig output with keys in the specified order.
+- `sh(sortedmulti(2,03acd484e2f0c7f65309ad178a9f559abde09796974c57e714c35f110dfc27ccbe,022f01e5e15cca351daff3843fb70f3c2f0a1bdd05e5af888a67784ef3e10a2a01))` describes a P2SH *2-of-2* multisig output with keys in the sorted lexicographically in the resulting redeemScript.
```
Comment (bitcoinhodler) [inline doc/descriptors.md]: s/keys in the sorted/keys sorted/

File: doc/descriptors.md [hunk 5]
```diff
@@ -37,12 +38,14 @@ Output descriptors currently support:
...
+- `wsh(sortedmulti(1,xpub661MyMwAqRbcFW31YEwpkMuc5THy2PSt5bDMsktWQcFF8syAmRUapSCGu8ED9W6oDMSgv6Zz8idoc4a6mr8BDzTJY47LJhkJ8UB7WEGuduB/1/0/*,xpub69H7F5d8KSRgmmdJg2KhpAK8SR3DjMwAdkxj3ZuxV27CprR9LgpeyGmXUbC6wb7ERfvrnKZjXoUmmDznezpbZb7ap6r1D3tgFxHmwMkQTPH/0/0/*))` describes a set of *1-of-2* P2WSH multisig outputs where the one multisig key is the *1/0/`i`* child of the first specified xpub and the other multisig key is the *0/0/`i`* child of the second specified xpub, and `i` is any number in a configurable range (`0-1000` by default). The order of public keys in the resulting witnessScripts is determined by the lexicographic order of the public keys at that index.
```
Comment (bitcoinhodler) [inline doc/descriptors.md]: s/where the one multisig key/where one multisig key/

File: doc/descriptors.md [hunk 6]
```diff
@@ -101,11 +105,12 @@ not contain "p2" for brevity.
 
 Several pieces of software use multi-signature (multisig) scripts based
 on Bitcoin's OP_CHECKMULTISIG opcode. To support these, we introduce the
-`multi(k,key_1,key_2,...,key_n)` function. It represents a *k-of-n*
+`multi(k,key_1,key_2,...,key_n)` and `sortedmulti(k,key_1,key_2,...,key_n)`
+functions. It represents a *k-of-n*
```
Comment (bitcoinhodler) [inline doc/descriptors.md]: s/It represents/They represent/ (?)

**Review by bitcoinhodler — CHANGES_REQUESTED** [review]
Love this.

File: src/script/descriptor.cpp (hunk 1 above)
Comment (achow101) [reply src/script/descriptor.cpp]: Done

File: src/script/descriptor.cpp (hunk 2 above)
Comment (achow101) [reply src/script/descriptor.cpp]: Done

File: doc/descriptors.md (hunk 4 above)
Comment (achow101) [reply doc/descriptors.md]: Done

File: doc/descriptors.md (hunk 5 above)
Comment (achow101) [reply doc/descriptors.md]: Done

File: doc/descriptors.md (hunk 6 above)
Comment (achow101) [reply doc/descriptors.md]: Done

File: test/functional/rpc_createmultisig.py [hunk 7]
```diff
@@ -72,6 +75,14 @@ def run_test(self):
             assert_equal(legacy_addr, node0.addmultisigaddress(2, keys, '', 'bech32')['address'])
             assert_equal(legacy_addr, node0.addmultisigaddress(2, keys, '', 'p2sh-segwit')['address'])
 
+        self.log.info('Testing sortedmulti descriptors with BIP 67 test vectors')
+        with open(os.path.join(os.path.dirname(os.path.realpath(__file__)), 'data/rpc_bip67.json'), encoding='utf-8') as f:
+            vectors = json.load(f)
+
+        for t in vectors:
+            key_str = ','.join(t['keys'])
+            assert_equal(self.nodes[0].deriveaddresses(descsum_create('sh(sortedmulti(2,' + key_str + '))'))[0], t['address'])
```
Comment (promag) [inline test/functional/rpc_createmultisig.py]: nit, in order to make this shorter:
```py
            keys = ','.join(t['keys'])
            desc = descsum_create('sh(sortedmulti(2,{}))'.format(keys))
            assert_equal(self.nodes[0].deriveaddresses(desc)[0], t['address'])

**Review by promag — COMMENTED** [review]
ACK 9214dd6f91, add release notes and call it a day.

File: test/functional/rpc_createmultisig.py (hunk 7 above)
Comment (achow101) [reply test/functional/rpc_createmultisig.py]: Done

**Comment by achow101** [comment]
Added a release note.

File: doc/release-notes-17056.md [hunk 8]
```diff
@@ -0,0 +1,4 @@
+Low-level RPC Changes
+===
+
+- A new descriptor type `sortedmulti(...)` has been added to support multisig scripst where the public keys are sorted lexicographically in the resulting script.
```
Comment (bitcoinhodler) [inline doc/release-notes-17056.md]: s/scripst/scripts/

Comment (achow101) [reply doc/release-notes-17056.md]: Done

**Comment by instagibbs** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/17056/commits/2638791a4f25394dc241367c933a2d27ed981197

**Review by Sjors — APPROVED** [review]
re-ACK 2638791a4f25394dc241367c933a2d27ed981197

File: test/functional/rpc_createmultisig.py [hunk 9]
```diff
@@ -72,6 +75,15 @@ def run_test(self):
             assert_equal(legacy_addr, node0.addmultisigaddress(2, keys, '', 'bech32')['address'])
             assert_equal(legacy_addr, node0.addmultisigaddress(2, keys, '', 'p2sh-segwit')['address'])
 
+        self.log.info('Testing sortedmulti descriptors with BIP 67 test vectors')
+        with open(os.path.join(os.path.dirname(os.path.realpath(__file__)), 'data/rpc_bip67.json'), encoding='utf-8') as f:
+            vectors = json.load(f)
+
+        for t in vectors:
+            key_str = ','.join(t['keys'])
+            desc = descsum_create('sh(sortedmulti(2,{}))'.format(key_str))
```
Comment (sipa) [inline test/functional/rpc_createmultisig.py]: You could check whether a `sh(multi(2,{}))` descriptor with the json's  "sorted_keys" gives the same result.

File: src/script/descriptor.cpp [hunk 10]
```diff
@@ -827,7 +836,10 @@ std::unique_ptr<DescriptorImpl> ParseScript(Span<const char>& sp, ParseScriptCon
         error = "Cannot have combo in non-top level";
         return nullptr;
     }
-    if (Func("multi", expr)) {
+    if (Func("sortedmulti", expr)) {
+        sorted_multi = true;
+    }
+    if (sorted_multi || Func("multi", expr)) {
```
Comment (sipa) [inline src/script/descriptor.cpp]: I very slightly prefer `if ((sorted_multi = Func("sortedmulti", expr)) || Func("multi", expr)) {`

**Comment by sipa** [comment]
ACK 2638791a4f25394dc241367c933a2d27ed981197. Agree on calling it `sortedmulti` as it's indeed more generic than bip67.

File: test/functional/rpc_createmultisig.py (hunk 9 above)
Comment (achow101) [reply test/functional/rpc_createmultisig.py]: Done

File: src/script/descriptor.cpp (hunk 10 above)
Comment (achow101) [reply src/script/descriptor.cpp]: Done

**Review by Sjors — APPROVED** [review]
re-ACK 4bb660be90a2811b53855bf1fd33a8dd9ba3db47

**Comment by instagibbs** [comment]
re-ACK https://github.com/bitcoin/bitcoin/pull/17056/commits/4bb660be90a2811b53855bf1fd33a8dd9ba3db47

**Pull request merged**